lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+
+  const size_t SrcOpsCount = getModuleOperations(SourceModule).size();
+  const size_t NewOpsCount = getModuleOperations(DestinationModule).size();
+  revng_assert(NewOpsCount == SrcOpsCount);
+
+  return DestinationModule;
+}
+
+static pipeline::Target makeTarget(const llvm::StringRef Name) {
+  return pipeline::Target(Name, TheKind);
+}
+
+static void makeExternal(mlir::FunctionOpInterface F) {
+  revng_assert(F->getNumRegions() == 1);
+  // A function is made external by clearing its region.
+  F->getRegion(0).getBlocks().clear();
+}
+
+// Remove any unused non-target symbols.
+static void pruneSymbols(mlir::ModuleOp Module) {
+  llvm::DenseSet<mlir::Operation *> UsedSymbols;
+
+  visit(Module, [&](mlir::FunctionOpInterface F) {
+    if (getFunctionName(F))
+      UsedSymbols.insert(F);
+
+    if (F.isExternal())
+      return;
+
+    if (const auto &Uses = mlir::SymbolTable::getSymbolUses(F)) {
+      for (const mlir::SymbolTable::SymbolUse &Use : *Uses) {
+        const auto &SymbolRef = Use.getSymbolRef();
+        revng_assert(SymbolRef.getNestedReferences().empty());
+
+        mlir::Operation *const
+          Symbol = mlir::SymbolTable::lookupSymbolIn(Module,
+                                                     SymbolRef
+                                                       .getRootReference());
+        revng_assert(Symbol);
+
+        UsedSymbols.insert(Symbol);
+      }
+    }
+  });
+
+  visit(Module, [&](mlir::SymbolOpInterface S) {
+    if (not UsedSymbols.contains(S.getOperation()))
+      S->erase();
+  });
+}
+
+const char MLIRContainer::ID = 0;
+
+void MLIRContainer::setModule(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule) {
+  revng_assert(NewModule);
+
+  // Make any non-target functions external.
+  visit(*NewModule, [&](mlir::FunctionOpInterface F) {
+    if (not F.isExternal() and not getFunctionName(F))
+      makeExternal(F);
+  });
+
+  pruneSymbols(*NewModule);
+  Module = std::move(NewModule);
+}
+
+std::unique_ptr<pipeline::ContainerBase>
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto NewContainer = std::make_unique<MLIRContainer>(name());
+  if (not getModuleBlock(*Module).empty()) {
+    OwningModuleRef NewModule(mlir::cast<mlir::ModuleOp>((*Module)->clone()));
+
+    bool RemovedSome = false;
+    visit(*NewModule, [&](mlir::FunctionOpInterface F) {
+      if (F.isExternal())
+        return;
+
+      const auto MaybeTargetName = getFunctionName(F);
+      if (not MaybeTargetName)
+        return;
+
+      if (not Filter.contains(makeTarget(*MaybeTargetName))) {
+        makeExternal(F);
+        RemovedSome = true;
+      }
+    });
+
+    if (RemovedSome)
+      pruneSymbols(*NewModule);
+
+    NewContainer->Context->appendDialectRegistry(Context->getDialectRegistry());
+    NewContainer->Module = cloneModuleInto(*NewModule, *NewContainer->Context);
+  }
+  return NewContainer;
+}
+
+void MLIRContainer::mergeBackImpl(MLIRContainer &&Container) {
```

Comment:
It would be nice if all relevant variables/arguments had a `Source` or `Destination` prefix. Same in `cloneFiltered`.